lib/logstash/plugin_mixins/scheduler.rb

```diff
@@ -0,0 +1,158 @@
+# encoding: utf-8
+
+require 'logstash/namespace'
+require 'logstash/plugin'
+require 'logstash/util/thread_safe_attributes'
+
+module LogStash
+  module PluginMixins
+    module Scheduler
+
+      extend LogStash::Util::ThreadSafeAttributes
+
+      ##
+      # @private
+      # @param base [Class]: a class that inherits `LogStash::Plugin`, typically one
+      #                      descending from one of the four plugin base classes
+      #                      (e.g., `LogStash::Inputs::Base`)
+      # @raise [ArgumentError]
+      # @return [void]
+      def self.included(base)
+        fail(ArgumentError, "`#{base}` must inherit LogStash::Plugin") unless base < LogStash::Plugin
+        instance_methods = base.instance_methods
+        base.send(:prepend, StopHook) if instance_methods.include?(:stop)
+        base.send(:prepend, CloseHook) if instance_methods.include?(:close)
+      end
+
+      # @private
+      module StopHook
+        def stop
+          release_scheduler! # wait till scheduler halts
+          super # plugin.stop
+        end
+      end
+      private_constant :StopHook
+
+      # @private
+      module CloseHook
+        def close
+          super # plugin.close
+          release_scheduler
+        end
+      end
+      private_constant :CloseHook
+
+      # def scheduler(); @_scheduler ||= new_scheduler({}) end
+      lazy_init_attr(:scheduler, variable: :@_scheduler) { start_scheduler({}) }
+
+      private
+
+      # Release jobs registered by the plugin from executing.
+      # This method executes from the plugin's #close method.
+      def release_scheduler
+        @_scheduler.release if @_scheduler
+      end
+
+      # Release jobs registered by the plugin from executing.
+      # This method executes from the plugin's #stop method.
+      # @note Might block until the scheduler operation completes!
+      def release_scheduler!
+        @_scheduler.release! if @_scheduler
+      end
+
+      # @param opts [Hash] scheduler options
+      # @return [SchedulerInterface] scheduler instance
+      def start_scheduler(opts, name: nil)
```

Comment:
Our only current non-private entry into this private `start_scheduler` is via the lazy-init `scheduler`, which passes neither `name` nor `opts`. Since this is a private method, I am okay with shipping as-is and leaving the cleanup/config for a later refactor. 

If/when we allow a consumer of this shared scheduler interface to configure the semantics of scheduling (such as execution overlap behaviour, error handling, etc.), I would like to do so in a way that isn't just passing Rufus' implementation details through.